### PR TITLE
DashboardScenes: Use behavior for library panels

### LINKED ISSUE #86681: Simplify library panel usage in dashboards
While migrating to the new Scenes-based dashboards we realized that the library panels migration caused a lot of code complexity and encumbered the architecture. It has resulted in a blocker for the library panel pane feature in panel edit ([PR1](https://github.com/grafana/grafana/pull/84493),[PR2](https://github.com/grafana/grafana/pull/84539)).

The way library panels are currently implemented adds a lot of complexity to the code that can be avoided with a refactor. Because of all this complexity, maintainability becomes a problem and so does adding new library panel related features in scenes dashboards. 

We need to simplify the work around library panels and around the `LibraryVizPanel`

### PR DESCRIPTION
**Pull request by mdvictor**
The current implementation of library panels in scenes is very complex to the point where it becomes very hard to maintain the code or add new features. This PR introduces a new implementation of library panels in the new architecture based on a custom `LibraryPanelBehavior`. This also fixes a bunch of issues found along the way during the migration, some of them existing in the current implementation.

Todo:
- [x] Migrate code to using behavior
- [x] Migrate tests
- [x] Remove LibraryVizPanel
- [x] Write tests for LibraryPanelBehavior
- [x] Fix unlink/replace in repeated lib panels
- [x] Adding a lib panel in empty saved dashboard through CTA didn't enter edit mode
- [x] Show panel loader until lib panel is fetched
- [x] Inspect panel json for lib panel doesn't bring up everything
- [x] Fix json model differences between old way of doing it
- [x] Show unsaved lib panel changes modal when leaving panel edit through breadcrumbs in panel edit
- [x] Fix height problem for repeated lib panels that causes them to not render

Fixes https://github.com/grafana/grafana/issues/86681

### CODE REVIEW

File: public/app/features/dashboard-scene/saving/DashboardPrompt.tsx [hunk 1]
```diff
@@ -38,29 +40,39 @@ export const DashboardPrompt = memo(({ dashboard }: DashboardPromptProps) => {
...
-    //     },
-    //     onDiscard: () => {
-    //       dispatch(discardPanelChanges());
-    //       moveToBlockedLocationAfterReactStateUpdate(location);
-    //       hideModal();
-    //     },
-    //     onDismiss: hideModal,
-    //   });
-    //   return false;
-    // }
```
Comment (torkelo) [inline public/app/features/dashboard-scene/saving/DashboardPrompt.tsx:63]: remove :)

**Review by torkelo — COMMENTED** [review]
Looking really good! Great work on this! 

Been testing a lot (and looking at the code). 

Doing a full page reload of a "view panel" is not working, troubleshooting / debugging this now

File: public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      pluginVersion: libPanelModel.pluginVersion,
+      displayMode: libPanelModel.transparent ? 'transparent' : undefined,
+      description: libPanelModel.description,
+      $data: createPanelDataProvider(libPanelModel),
+    };
+
+    this.setState({ _loadedPanel: libPanel, isLoaded: true, name: libPanel.name, title: libPanelModel.title });
+
+    const clone = vizPanel.clone(vizPanelState);
+    gridItem.setState({ body: clone });
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: My biggest problem with this behavior approach which I discovered fixing issues in https://github.com/grafana/grafana/pull/92846/ 

is this setState. Feels a bit broken that the behavior replaces it's parent instance (and clones itself), by going to it's parent's parent (the DashboardGridItem) and updates it's state. 

I think we could get it work without updating the grid item state and creating a VizPanel instance

if we did: 
```ts
    vizPanel.setState(vizPanelState);
    vizPanel.changePluginType(libPanelModel.type, vizPanelState.options, vizPanelState.fieldConfig);
```

It seems to work. For panel edit we would still need to wait (as doing this after we open panel edit could have issues, but might work)

Comment (torkelo) [reply public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: I explored doing that and could then get much simpler fix for the view panel / edit panel (from full page reload)

https://github.com/grafana/grafana/pull/92850

Comment (mdvictor) [reply public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: @torkelo I just went through this again, just to make sure I understand: What isn't working properly in this PR on page reload in edit/view panel? I just retested this and it loads in both view/edit on page reload. What am I missing?

**Comment by mdvictor** [comment]
Think I found another issue: after a edit panel page reload if we try to just return to the dashboard without any changes the "Unsaved changes" prompt will pop up even though nothing is changed, or shouldn't at least. Even though I don't do anything in the UI there are differences in the vizPanel state due to lib panel loading I assume? E.g.: `pluginVersion` is missing (and probably others pulled from the libPanel that are not stored in the json?) so that sets the panel dirty and triggers the prompt.

**Comment by torkelo** [comment]
@mdvictor I guess that is also true in main, so I would not worry about it for this change

File: public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: @mdvictor eh, that is super strange, a simple lib panel added to a dashboard, then save. go into view panel, do a full page reload, see not correct queries not viz. 

Same issue in panel edit, and the explanation is super easy to understand in view panel and panel edit the VizPanel parent is not g grid item so no replacement of the VizPanel with the loaded lib panel happens.

[testing---Dashboards---Grafana.webm](https://github.com/user-attachments/assets/95936e36-bf8a-40f3-9739-db0733e0dee5)

Comment (torkelo) [reply public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: @mdvictor I think you must have tested with a repeated lib panel, and the reason it works there is that in that case we wait for the repeat process to complete before going into view / edit, and that wait gives the behavior time to load and replace the viz panel (but depends on timing of the load lib panel API call I guess, if it was slow we would still see bug there)

but just test with a lib panel that is not repeated

Comment (mdvictor) [reply public/app/features/dashboard-scene/scene/LibraryPanelBehavior.tsx:65]: Ahh yes, I see it now, thank you!

**Review by torkelo — APPROVED** [review]

**Pull request merged**